text/0082-entity-tags.md

```diff
@@ -0,0 +1,251 @@
+# Entity tags with dedicated syntax and semantics
+
+## Related issues and PRs
+
+* Reference Issues: [#305](https://github.com/cedar-policy/cedar/issues/305)
+* Supersedes [RFC 68](https://github.com/cedar-policy/rfcs/pull/68)
+* Implementation PR(s): (leave this empty)
+
+## Timeline
+
+* Started: 2024-09-11
+* Accepted: TBD
+* Landed: TBD
+* Released: TBD
+
+## Summary
+
+This RFC proposes to extend the Cedar language, type system, and symbolic analysis to include full-featured *entity tags* for entity types. Tags are a mechanism used by cloud services to attach key-value pairs to resources. Cedar will allow them to be attached to any entities (not just resources).
+
+For evaluation purposes, entity tags are accessed using the methods `expr1.hasTag(expr2)` and `expr1.getTag(expr2)`, where `expr1` evaluates to an entity and `expr2` to a string.  These are the only ways to access the tags on an entity.  The policy syntax, semantics, type system, and symbolic analysis are extended to support these two methods.  The schema syntax and the entity JSON format are extended to support defining entity tags and their types.
+
+Entity tags strictly increase the expressiveness of Cedar because they allow *computed keys*, unlike records and entities. Specifically, tag keys need not be literals, so it is possible to write `resource.hasTag(context.tag)` and `resource.getTag(context.tag)`.
+
+This proposal is backward-compatible in the sense that existing polices, schemas, entities, etc., do not need to change to accommodate the addition of tags to the language.
+
+This RFC has gone through several revisions, with different revisions proposing alternative designs. The alternatives are discussed at the end of this document.
+
+### Basic example
+
+Here is a schema defining two entities, each of which contains entity tags.
+
+```
+entity User = {
+  jobLevel: Long,
+} tags Set<String>;
+entity Document = {
+  owner: User,
+} tags Set<String>;
+```
+
+The `User` and `Document` entities each have entity tags, denoted by `tags Set<String>`, implementing tags whose values are sets of strings.  The declaration `tags Set<String>` indicates an unspecified number of optional tags that all have values of type `Set<String>`.  (Of course, any type could be written, for instance `Long`, not just `Set<String>`.) Note that the set of tag (keys) does not have to be defined in the schema, but the type of all tag values needs to be the same (here, `Set<String>`).  (You cannot have tag `foo` have value `2`, and tag `bar` have value `"spam"`.)  Note also that there is no way to name the key-value map representing tags; you cannot treat it as a Cedar record or even compare it for equality to the tag-map on a different entity; you can only compare the values of individual keys.
```

Comment:
I'm not sure I fully grok the current capabilities of Cedar + this proposal, but I do have a real-world example I'd want to walk through.

In Kubernetes, admission gets both the requesting user and the resource being acted on. I'd really like to be able to implement some sort of ABAC, but the the tags I'd be working with are slightly different shapes. Every Kubernetes resource has key/value (`String`/`String`) set of labels. 

The User structure has extra info in the form of `String`/`Set<String>`. Would it be as simple as using existing `.contains()` functions in policy to implement ABAC?

```
permit (
  principal is User,
  action == Action::"writeDoc",
  resource is Document)
when {
  (principal.hasTag("my.tld/myorg") &&
   resource.hasTag("my.tld/myorg") &&
   principal.getTag("my.tld/myorg").contains(resource.getTag("my.tld/myorg"))
};
```

Reply:
If the resource Entity Type is defined to have `tags String` and the principal Entity Type is defined to have `tags Set<String>`, then the example policy is type-correct and will work as expected.

Here, I would expect the policy to check if`resource.getTag("my.tld/myorg")`, which is a String, is included in the set of Strings `principal.getTag("my.tld/myorg")`.